src/dexer.rs

```diff
@@ -1,10 +1,255 @@
+    release: bool,
+
+    /// Specify the minimum Android API level you want the output DEX files to support.
+    android_min_api: Option<u32>,
+
+    /// Disable Java 8 language features. Use this flag only if you don't intend to compile
+    /// Java bytecode that uses language features introduced in Java 8.
+    no_desugaring: bool,
+
+    /// Specify the path to the `android.jar` of your Android SDK.
+    android_jar_path: Option<PathBuf>,
+
+    /// Specify classpath resources that `d8` may require to compile your project's DEX files.
+    class_paths: Vec<OsString>,
+
+    /// Specify the desired path for the DEX output. By default, `d8` outputs the DEX file(s)
+    /// in the current working directory.
+    out_dir: Option<OsString>,
+
+    /// Specifies paths to compiled Java bytecodes that you want to convert into DEX bytecode.
+    /// The input bytecode can be in any combination of `*.class` files or containers, such as
+    /// JAR, APK, or ZIP files.
+    files: Vec<OsString>,
+}
+
+impl Dexer {
+    /// Creates a new `Dexer` instance with default values,
+    /// which can be further customized using the builder methods.
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Executes the `java` command based on this `Dexer` instance.
+    pub fn run(&self) -> std::io::Result<ExitStatus> {
+        self.command()?.status()
+    }
+
+    /// Returns a [`Command`] based on this `Dexer` instance
+    /// that can be inspected or customized before being executed.
+    pub fn command(&self) -> std::io::Result<Command> {
+        let mut d8_run = JavaRun::new();
+        
+        if let Some(java_home) = &self.java_home {
+            d8_run.java_home(java_home);
+        }
+
+        let d8_jar_path = self.android_d8_jar_path
+            .clone()
+            .and_then(PathExt::path_if_exists)
+            .or_else(|| env_paths::android_d8_jar(None))
+            .ok_or_else(|| std::io::Error::other(
+                "d8.jar not provided, and could not be auto-discovered."
+            ))?;
+
+        d8_run.class_path(d8_jar_path)
+            .main_class("com.android.tools.r8.D8");
+
+        if self.release {
+            d8_run.arg("--release");
+        }
+
+        if let Some(min_api) = self.android_min_api {
+            d8_run.arg("--min-api").arg(min_api.to_string());
+        }
+
+        if self.no_desugaring {
+            d8_run.arg("--no-desugaring");
+        } else {
+            let android_jar_path = self.android_jar_path
+                .clone()
+                .and_then(PathExt::path_if_exists)
+                .or_else(|| env_paths::android_jar(None))
+                .ok_or_else(|| std::io::Error::other(
+                    "android.jar not provided, and could not be auto-discovered."
+                ))?;
+            d8_run.arg("--lib").arg(android_jar_path);
+
+            for class_path in &self.class_paths {
+                d8_run.arg("--classpath").arg(class_path);
+            }
```

Comment:
can you add a comment explaining why you only do these steps if `no_desugaring` is false? They seem sort of unrelated.

Perhaps it's obvious but I'm not deeply acquainted with dex tools in general.